apps/omg_api/lib/state/core.ex

```diff
@@ -168,98 +164,105 @@ defmodule OMG.API.State.Core do
-    end
+    if no_utxo_from_future_block, do: :ok, else: {:error, :input_utxo_ahead_of_state}
   end
 
-  defp correct_input_in_position?(
-         2,
-         state,
-         %Transaction{blknum2: blknum, txindex2: txindex, oindex2: oindex, cur12: spent_cur},
-         spender
+  defp inputs_belong_to_spenders?(
+         utxos,
+         %Transaction.Recovered{
+           signed_tx: %Transaction.Signed{raw_tx: raw_tx}
+         } = recovered_tx
        ) do
-    with :ok <- utxo_not_from_the_future_block?(state, blknum) do
-      check_utxo_and_extract_amount(state, Utxo.position(blknum, txindex, oindex), spender, spent_cur)
+    inputs = Transaction.get_inputs(raw_tx)
+
+    with {:ok, input_utxos} <- get_input_utxos(utxos, inputs),
+         input_utxos_owners <- Enum.map(input_utxos, fn %{owner: owner} -> owner end),
+         :ok <- Transaction.Recovered.all_spenders_authorized?(recovered_tx, input_utxos_owners) do
+      {:ok, input_utxos}
     end
   end
 
-  defp check_utxo_and_extract_amount(%Core{utxos: utxos}, position, spender, spent_cur) do
-    with {:ok, %Utxo{owner: owner, currency: cur, amount: owner_has}} <- get_utxo(utxos, position),
-         :ok <- is_spender?(owner, spender),
-         :ok <- same_currency?(cur, spent_cur),
-         do: {:ok, owner_has}
+  defp get_input_utxos(utxos, inputs) do
+    inputs
+    |> Enum.filter(fn Utxo.position(blknum, _, _) -> blknum != 0 end)
+    |> Enum.reduce({:ok, []}, fn input, acc -> get_utxos(utxos, input, acc) end)
   end
 
-  defp utxo_not_from_the_future_block?(%__MODULE__{height: blknum}, input_blknum) do
-    if blknum >= input_blknum do
-      :ok
-    else
-      {:error, :input_utxo_ahead_of_state}
-    end
-  end
+  defp get_utxos(_, _, {:error, _} = err), do: err
 
-  defp get_utxo(utxos, position) do
+  defp get_utxos(utxos, position, {:ok, acc}) do
     case Map.get(utxos, position) do
       nil -> {:error, :utxo_not_found}
-      found -> {:ok, found}
+      found -> {:ok, [found | acc]}
     end
   end
 
-  defp is_spender?(owner, spender) do
-    if owner == spender, do: :ok, else: {:error, :incorrect_spender}
+  defp get_amounts_by_currency(utxos) do
+    utxos
+    |> Enum.group_by(fn %{currency: currency} -> currency end, fn %{amount: amount} -> amount end)
+    |> Enum.map(fn {currency, amounts} -> {currency, Enum.sum(amounts)} end)
+    |> Map.new()
   end
 
-  defp same_currency?(utxo_currency, spent_currency) do
-    if utxo_currency == spent_currency, do: :ok, else: {:error, :incorrect_currency}
+  # fee is implicit - it's the difference between funds owned and spend
+  defp amounts_add_up?(input_amounts, output_amounts, fees) do
+    outputs_covered =
+      for {output_currency, output_amount} <- Map.to_list(output_amounts) do
+        fee = Map.get(fees, output_currency, 0)
+        input_amount = Map.get(input_amounts, output_currency, 0)
+        input_amount >= fee + output_amount
+      end
+      |> Enum.all?()
+
+    fees_covered =
```

Comment:
ah I see the point with fees better now. This is a bit that's a little underdefined on the requirements end.

In this implementation, the fee must be paid for every currency that pops up in input or output.

That's probably not what we intend - most likely we want at least one currency of input to cover the fee. This way the user chooses which currency to pay with.

But we can amend it (and document!) in a sequel PR to not delay this. Please loop in @kasima and/or @whoisjeremylam to confirm on the desired behavior.

Reply:
this still doesn't behave as expected I think. 1/ it requires all fees to be paid up for all inputs 2/ it additionally has a `fees_covered` condition which is a list that's treated as a bool, which checks additionally, if there any excess inputs also cover the fee.

Are there `Core` tests that cover these behaviors so we can see how it works?